doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+requires the use of hugepages, as it does on Cray X\* systems.
+
+
+Single-ddata with Separate Arrays
+'''''''''''''''''''''''''''''''''
+
+**Synopsis**: as above, but allocate and localize big arrays separately
+
+**Performance snapshot**
+
+    ==========================  ====
+    execution/storage affinity  fair
+    access speed                best
+    balance                     fair
+    leverage locality           fair
+    network cooperation         fair
+    ==========================  ====
+
+
+Description
+...........
+
+The most serious failing of the ordinary single-ddata model is its poor
+cooperation with NIC-registered memory, whether or not that memory has
+NUMA locality.  This can be improved by adopting an idea described
+during the discussions leading up to this document.  That is to allocate
+NUMA-localized arrays individually from memory outside any existing NIC
+registration, and do the localization, registration, and registration
+broadcast to remote nodes after allocating.  This would avoid the
+re-registration problem created if we allocated from NIC-registered
+memory and then wanted to change its locality.
+
+
+Analysis
+........
+
+We would not necessarily need to invalidate such array registrations
+when the arrays were freed, because references to freed arrays are
+nonconforming in Chapel.  We would nevertheless need to invalidate such
+a registration if, after one array had been allocated and freed, a later
+allocation happened to overlap the same storage.  We would have to
+ensure that any other node referencing the new array had received the
+new registration before doing so.  This is no problem technically: the
+allocating node can broadcast the new registration info and wait for
+acknowledgment from all remotes before continuing.  But it does mean a
+one-to-all round-trip communication at allocation and registration time,
+or even effectively all-to-all communication if all nodes are behaving
+in synchrony.  The impact on overall application performance would
+depend on how often such updates were needed.
+
+A brief search for applications or mini-apps which allocate and free
+many NUMA-sensitive arrays turned up one that looks like this:
+
+.. code-block:: chapel
+
+    for layer in 1..numLayers {
+       var curLayer: [0..#nextLayer.size] node;
+       curLayer = nextLayer;
+       resize(nextLayer, n);
+
+       // I *think* this could be converted to a forall with some effort
+       for v in curLayer {
+           var idx: int = complicatedExpensiveComputation(v);
+           nextLayer[idx] = someValue;
+       }
+    }
+
+Note that if either the for-stmt cannot be converted to a forall-stmt or
+``complicatedExpensiveComputation()`` really is expensive then the
+``nextLayer`` references in the kernel may not really be NUMA-sensitive.
+But for now this is is all we have seen that might be sensitive to the
+cost of array-by-array registration as described here.
```

Comment:
Yeah.  But that might get complicated depending on where (which cache) things were spatially cached in. If they're in an L1 cache at or near the processor that would hide the effect of getting the NUMAness wrong, but if they were in an L3 cache on the NUMA domain it wouldn't.  I'd prefer to avoid that level of detail.  I did do a lot of rephrasing on this part, so see how you like the next version.